odl/contrib/pytorch/layer.py

```diff
@@ -0,0 +1,242 @@
+        >>> matrix_op = TorchOperator(op)
+        >>> matrix_op.operator is op
+        True
+        """
+        self.operator = operator
+
+    def forward(self, input):
+        """Evaluate forward pass on the input.
+
+        Parameters
+        ----------
+        input : `torch.tensor._TensorBase`
+            Point at which to evaluate the operator.
+
+        Returns
+        -------
+        result : `torch.autograd.variable.Variable`
+            Variable holding the result of the evaluation.
+
+        Examples
+        --------
+        Perform a matrix multiplication:
+
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TorchOperator(op)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x)
+        >>> matrix_op(x_var)
+        Variable containing:
+         4
+         5
+        [torch.DoubleTensor of size 2]
+
+        Evaluate a functional, i.e., an operator with scalar output:
+
+        >>> space = odl.rn(3)
+        >>> functional = odl.solvers.L2NormSquared(space)
+        >>> func_op = TorchOperator(functional)
+        >>> x_elem = space.element([1, 2, 3])
+        >>> x = torch.from_numpy(x_elem.asarray())
+        >>> x_var = torch.autograd.Variable(x)
+        >>> func_op(x_var)
+        Variable containing:
+         14
+        [torch.DoubleTensor of size 1]
+        """
+        if not self.operator.is_linear:
+            # Only needed for nonlinear operators
+            self.save_for_backward(input)
+
+        input_arr = input.numpy()
+        if all(s == 0 for s in input_arr.strides):
+            # TODO: remove when Numpy issue #9165 is fixed
+            # https://github.com/numpy/numpy/pull/9177
+            input_arr = input_arr.copy()
+
+        op_result = self.operator(input_arr)
+        return torch.from_numpy(np.array(op_result, copy=False, ndmin=1))
+
+    def backward(self, grad_output):
+        """Apply the adjoint of the Jacobian at ``grad_output``.
+
+        This method is usually not called directly but as a part of the
+        ``cost.backward()`` pass of a backpropagation step.
+
+        Parameters
+        ----------
+        grad_output : `torch.tensor._TensorBase`
+            Tensor to which the Jacobian should be applied. See Notes
+            for details.
+
+        Returns
+        -------
+        result : `torch.autograd.variable.Variable`
+            Variable holding the result of applying the Jacobian to
+            ``grad_output``. See Notes for details.
+
+        Examples
+        --------
+        Compute the Jacobian adjoint of the matrix operator, which is the
+        operator of the transposed matrix. We compose with the ``sum``
+        functional to be able to evaluate ``grad``:
+
+        >>> space = odl.rn(3)
```

Comment:
This stuff is DENSE for a doctest. Is there no way to make a more simple example?

Reply:
I'll see what I can do.

Reply:
One shortcut: `x = torch.DoubleTensor([1, 2, 3])` instead of the two current lines.